README.md
# Fachwerk

Fachwerk is a Javascript framework for creating interactive learning materials in the browser.

Content can be authored in a Markdown format, with custom additions such as dynamic layouts, interactivity and wide range of HTML-like components.

Fachwerk is based on [VueJS](https://vuejs.org) and the latest browser technologies such as Javascript imports / exports and CSS variables. It provides an easy onboarding without any complex tooling or setup. Yes, this means 🦄 **there is no build step**.

## Links

🏰 <a href="https://designstem.github.io/fachwerk">Homepage</a>

📖 <a href="https://designstem.github.io/fachwerk/docs">Documentation</a>

🚀 <a href="https://designstem.github.io/fachwerk/docs/#/start-a-new-project"> Quickstart guide</a>

## Custom start

Easiest way to have a custom Fachwerk application is to create a three files in your machine

##### 1. index.html

```html
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <link
      rel="stylesheet"
      href="https://designstem.github.io/fachwerk/fachwerk.css"
    />
  </head>
  <body>
    <div id="fachwerk"></div>
    <script type="module" src="./index.js"></script>
  </body>
</html>
```

##### 2. index.js

```js
import { fachwerk } from "https://designstem.github.io/fachwerk/fachwerk.js";
fachwerk();
```

##### 3. index.md

```md
# Hello 🌏

<f-scene grid>
  <f-circle />
</f-scene>
```

To run the project in your browser, you have to host the files in your local server. Easiest way to do it is to install [Web Server for Chrome](https://chrome.google.com/webstore/detail/web-server-for-chrome/ofhbbkphhbklhfoeikjpcbhemlocgigb?hl=en) and point it to your project folder. You can also use the commandline tools of your choice.

## Customizing the framework

In many cases the framework needs additional customization. You can start with small tweaks, add custom components or use set up your project in whole custom way.

### Custom configuration

Fachwerk has several setup options you can pass in as an object:

```
fachwerk({ el: "#fachwerk" })
```

| key          | default        | options                                | description                                          |
| ------------ | -------------- | -------------------------------------- | ---------------------------------------------------- |
| `el`         | `"#fachwerk"`  |                                        | HTML tag id the content goes into                    |
| `src`        | `"./index.md"` |                                        | Path of the Markdown file, local path or web address |
| `title`      | `"Fachwerk"`   |                                        | Title of the project. Used for local storage ID      |
| `edit`       | `"hide"`       | `"none"` `"hide"` `"show"`             | Show integrated live editor?                         |
| `menu`       | `"hide"`       | `"none"` `"hide"` `"show"`             | Show navigation menu?                                |
| `type`       | `"document"`   | `"slides"` `"document"`                | Presentation mode                                    |
| `typebutton` | `"show"`       | `"none"` `"show"`                      | Show presentation mode switcher?                     |
| `home`       | `""`           |                                        | Home navigation URL, local file path or web address  |
| `theme`      | `"light"`      | `"light"` `"dark"` `"blue"` `"yellow"` | Global theme                                         |
| `style`      | `{}`           |                                        | Add custom styles                                    |
| `components` | `{}`           |                                        | Add custom components                                |
| `utils`      | `{}`           |                                        | Add custom utilities                                 |

### Custom styles

Fachwerk supports custom style overrides, they are especially helpful setting global CSS variables.

Here is how to do it:

##### index.js

```js
fachwerk({ styles: { "--content-padding": "var(--base10)" } });
```

### Custom components

Fachwerk builtin components inside Markdown files are very powerful, but they can can be to long and unwieldy to be handled inside content.

The solution is to extract complex components to separate component files. They can still communicate with builtin components and document with global state and events.

Here is how to do it:

##### index.js

```js
import { fachwerk } from "https://designstem.github.io/fachwerk/fachwerk.js";

import CustomComponent from "./CustomComponent.js";

fachwerk({ components: { CustomComponent } });
```

##### CustomComponent.js

```js
import { utils } from "https://designstem.github.io/fachwerk/fachwerk.js";

export default {
  methods: { ...utils },
  template: `<div>Hey, I am a custom component</div>`
};
```

##### index.md

```md
# Hallo Welt

<custom-component />
```

See VueJS [component documentation](https://vuejs.org/v2/guide/components.html) for more details.

### Custom utilities

Sometimes it might be useful to have your own set of utility functions to be available for templates, here is how to add them to framework:

##### utils.js

```js
export const add = (a, b) => a + b;
```

##### index.js

```js
import { fachwerk } from "https://designstem.github.io/fachwerk/fachwerk.js";

import * as utils from "./utils.js";

fachwerk({ utils });
```

### Custom project

There are cases you want a full control how the framework is set up. Just replace your index.js with the following:

##### index.js

```js
import {
  Vue,
  components,
  utils
} from "https://designstem.github.io/fachwerk/fachwerk.js";

// Register components globally

for (const name in components) {
  Vue.component(name, components[name]);
}

// If you have custom components,
// install them here:
//
// import CustomComponent from "./CustomComponent.js";
// Vue.component("custom-component", CustomComponent);

// Set up a global event bus

Vue.prototype.$global = new Vue({ data: { state: {} } });

// Initialize VueJS

new Vue({
  // HTML tag id the content goes into

  el: "#fachwerk",

  methods: {
    // Allow utils to be used in templates

    ...utils

    // Custom methods go here
  },

  // Reactive data will be here
  // Note that it will be only accessible
  // in the template, not inside Markdown
  // Use get() / set() helpers to
  // pass values to index.md content

  data: {},

  // In the main template fetch and
  // display index.md. This template
  // can be totally custom and can
  // mix and match Fachwerk and custom
  // components and utilities.

  template: `
  <f-fetch src="./index.md" v-slot="{ value: content }">
    <f-content-editor :content="content" />
  </f-fetch>
  `
});
```


RELEASES.md
# Release notes

### 0.0.9 / Mon Nov 4, 2019

#### Rewritten layout, editor and preview components

- There is a new version of layout / integrated editor. This includes some breaking changes in `fachwerk()` config:

`config.editor` is now `config.edit`

`config.menu` accepts values `none | hide | show`

`config.pager`, `config.header` and `config.footer` are no longer supported.

#### Docs

- New set of tutorials, featuring quickstart, drawing with math, making generative art and more

- Removed old outdated tutorials

#### New features

- New `<f-content-example>` component to provide interactive code samples.

- New `<f-triangle>` component.

- New `<f-polargrid>` component and `<f-scene polargrid>` option.

- New `<f-mouse>` (experimental) component.

- `| backgroundposition: 50% 50%` option in slides to position background images.

- `--mobile-cols`, `--mobile-rows` and `--mobile-gap` options on `<div class="grid">`

- New `--area: a b'` option on `<div class="grid">` (children have to have `<div style="grid-area='a'">` and `<div style="grid-area='b'">` set).

- `<f-footer>` has now minimal styling and no padding by default.

#### Bug fixes

- `FSidebar` and other components default to full width on mobile

### 0.0.8 / Mon 15 Jul 2018

#### New features

- New `<f-toggle>` component

- `<f-scene>`, `<f-scene3>`, `<f-artboard>` and `<f-canvas>` now have `download` prop that allows download the scene as SVG or PNG

- `<f-scene>`, `<f-scene3>` and `<f-artboard>` have `responsive` tag to scale the scene to the container width

- `<f-scene3>` can have `static` drop that improves the performance when you do not need live editing or animations

- New pattern component `<f-grid-pattern3>`

- `<f-math>` now has `inline` prop for inline math and various color props for math coloring

- Slides can now have `style:` setting where you can set the CSS styles

- Slides `background:` setting can now be either a CSS variable color or an image

- New utility functions: `object2array()`, `camelcase()`

- `<f-scene3>` scenes now support retina screens

### 0.0.7 / Mon 10 Jun 2018

- Adding `<f-piano>` onscreen keyboard component

- Adding `chords()` and `chord()` music utility functions

- Adding `intersection()` array utility function

- Adding `colors()` function to return list of framework color names

- In `<f-fetch>`, `src` prop can be an array and load multiple files

#### Depreciations

- `url` prop is replaced with `scr` prop in `<f-fetch>`

### 0.0.6 / Mon 3 Jun 2018

#### New features

- Orthogonal / isometric rendering for 3d scenes. Can be set with `<f-scene3 isometric>` prop.

- `f-extrude3` component to create extruded geometry from 2D polyline.

- Experimental audio and MIDI components `f-synth`, `f-sequencer`, `f-drum`, `f-midi-info`, `f-midi-in`, `f-midi-out` and beginnings of music making tutorial.

#### Improvements

- Code editor now wraps long URLs in code and there is a `--content-editor-min-width` CSS variable to control minimum width.

- Ordered lists are now styled with bullets.

- `f-mirror-x` and `f-mirror-y` now accept `step` prop to control distance between original and mirrored element.

#### Depreciations

- `<f-scene3 renderer="webgl" />` prop has been replaced with `<f-scene3 webgl />`

### 0.0.5 / Mon 6 May 2018

#### Improvements

- Experimental `f-websocket` component, allowing bidirectional realtime communication

- Adding transparent color, accessible via `color('transparent')` function and `var(--transparent)` CSS variable

- Slides have now `| rows:` and `| cols:` parameters, allowing setting CSS grid rows and cols.

- Slide layouts are now responsive (ongoing work)

- `f-video` Youtube URL accepts `?start=` parameter for playing from specified timestamp

- Icon colors are set by `--icon-stroke` and `--icon-fill` CSS variables

#### Docs

- New docs structure, pages are consolidated by topic

- New introduction docs for layout, navigation and interaction

### 0.0.4 / Mon 29 April 2018

#### Improvements

- In scenario slides editor is hidden by default. You can control it in `index.js` by setting

  fachwerk({ editor: 'hide'})
  fachwerk({ editor: 'show'})
  fachwerk({ editor: 'none'})

- `<f-notes>` accepts `title` attribute and defaults to "Teacher notes"

- Initial improvements to the docs menu structure

- New icons for internal use: `<f-people>`, `<f-clock>`, `<f-tools>` (last one is placeholder for now)

### 0.0.3 / Mon 22 April 2018

#### Improvements

- New `polarxy(angle, radius)` function, returns [x,y] coordinates, works nicely with `:position` prop on geometry components

- Added `send('edit')` event that toggles edit mode for `<f-content-editor />`

### 0.0.2 / Mon 15 April 2018

#### Improvements

- New top navigaton for menu, editor, notes and paging buttons. This work is ongoing

- Slides can be navigated with right / left buttons. alt + right / alt + left should be used when on-page editor is open

- New content icons: `<f-activity-icon />`, `<f-fact-icon />`, `<f-note-icon />` and `<f-vr-icon />`

- New navigation icons (mostly used internally): `<f-menu-icon />`, `<f-rightarrow-icon />`, `<f-leftarrow-icon />`, `<f-close-icon />`

- New math functions: `distance()` to calculate distance between two points and `linepoint()` to calculate single point coordinates on the line

- Added `goto(id)` utility function. Use it everywhere you used `send('goto', id)`

- Global CSS variable `--transition-duration` for setting CSS transition speeds

- `<f-animation>` now supports `delay` parameter

- Fading transitions on slides and sidebar / notes / menu appearances

- Currently active slide will be remembered on page reload

#### Experimental functionality

- Initial work for having sections (subgroup of slides) in content. Documentation will be added on upcoming releases

- Experimental `<f-section-card>` component for setting up menus. Documentation will be added on upcoming releases


index2.js
import * as components from "./framework.js";

for (const name in components) {
  Vue.component(name, components[name]);
}

import { sortedComponents } from "./framework.js";

import Colors from "./homepage/Colors.js";
import Styles from "./homepage/Styles.js";
import ComponentRow from "./homepage/ComponentRow.js";
import Utils from "./homepage/Utils.js";
import Faq from "./homepage/Faq.js";

// const PropsTable = {
//   props: { props: { default: {}, type: Number } },
//   computed: {
//     propRows() {
//       return Object.entries(this.props).map(p => ({
//         Name: `<code>${p[0]}</code>`,
//         Default: `<code>${p[1].default}</code>`,
//         Type: `<code>${
//           Array.isArray(p[1].default) ? "array" : typeof p[1].default
//         }</code>`
//       }));
//     }
//   },
//   template: `
//     <f-table :rows="propRows" style="--lightblue: transparent" />
//   `
// };

// const ComponentRow = {
//   props: {
//     component: { default: {}, type: Object }
//   },
//   methods: { kebabCase },
//   components: { PropsTable },
//   template: `
//     <div>
//       <h3>&lt;{{ kebabCase(component.name) }}&gt;</h3>
//       <div class="grid" style="--cols: 1fr 3fr">
//         <div>
//           <markdown v-if="component.description" :content="component.description" />
//           <div style="max-height: calc(var(--base) * 50); overflow: auto;">
//             <h5>Props</h5>
//             <PropsTable :props="component.props" />
//           </div>
//         </div>
//         <div>
//           <f-content-editor
//             v-if="component.example"
//             :content="component.example.trim()"
//             :autosave-id="component.name"
//             style="min-height: calc(var(--base) * 15)"
//           ><f-content-document slot-scope="{content}" :content="content" /></f-content-editor> 
//         </div>
//       </div>
//       <br><br>
      
//     </div>
//   `
// };

// const Colors = {
//   mixins: [Css],
//   template: `
//   <div class="colors">
//     <template v-for="c in [
//     'purple', 
//     'red',
//     'yellow',

//     'darkblue',
//     'blue',
//     'lightblue',

//     'darkestgray',
//     'darkergray',
//     'darkgray',

//     'gray',
//     'lightgray', 
//     'lightergray', 
//   ]">
//     <div :style="{
//       background: 'var(--' + c + ')',
//       borderRadius: '1000px'
//     }"></div>
//       <code style="
//         color: var(--secondary);
//         background: none;
//         margin: 0.5rem 0 0 0;
//         line-height: 1.4rem;"
//       >
//         var(--{{ c }})
//         <br>
//         color('{{ c }}')
//       </code>
//     </template>
// </div>
// `,
// css: `
// .colors {
//   display: grid;
//   grid-template-columns: 4rem 5fr 4rem 5fr 4rem 5fr;
//   grid-template-rows: repeat(4, 4rem);
//   grid-gap: 0.5rem;
// }
// @media (max-width: 900px) {
//   .colors {
//     grid-template-columns: 4rem 1fr;
//     grid-template-rows: repeat(12, 4rem);
//   }
// }
// `
// }

// const Styles = {
//   template: `
//   <div class="grid" style="--cols: 1fr 1fr 1fr;">
//   <div>
//     <dd>h2 .bullet</dd>
//     <h2><span class="bullet">2</span>Here is the smaller subtitle</h2>
//     <dd>h3 .bullet</dd>
//     <h3><span class="bullet">3</span>Here is the even smaller subtitle</h3>
//     <dd>h3 .bullet</dd>
//     <h4><span class="bullet">4</span>Here is the even more smaller subtitle</h4>
//     <dd>.intro</dd>
//     <div class="intro">Here is a intro with <a href="https://en.wikipedia.org/wiki/Johannes_Gutenberg">some
//         links</a> in it. Actually, it can be even longer if needed</div>
//     <dd>p, p b, p i</dd>
//     <p>This is just a regular paragraph text with some <b>basic</b> <i>formatting</i> if one so desires.</p>
//     <dd>p small, p small b, p small i</dd>
//     <p><small>This is just a small regular text wish some <b>basic</b> <i>formatting</i> if one so desires.</small></p>
//   </div>

//   <div>
//     <dd>.text, .text b, .text i</dd>
//     <div class="text">
//       <p>Normal polygons are
//         <b>polygons</b> with all sides and angles equal. Are formed by triangles, normally
//         <i>isosceles</i>.
//       </p>
//       <dd>.text p, .text code, .text a</dd>
//       <p>For example <code>6px</code> triangles connected together form a regular polygon. Normal polygons are characterized by the number of edges or sides.</p>
//       <p>Normal polygons are characterized by the number of edges or sides. For example 6px triangles connected together form a regular polygon.</p>
      
//       <dd>.text p small</dd>
//       <small>
//           For example 6px triangles connected together form a regular polygon. Normal polygons are characterized
//           by the number of edges or sides.
//       </small>
//     </div>

//     <dd>table</dd>
    
//     <table>
//       <thead>
//         <th>First col</th>
//         <th>Second col</th>
//       </thead>
//       <tbody>
//         <tr>
//           <td>First col</td>
//           <td>Second col</td>
//         </tr>
//         <tr>
//             <td>First col</td>
//             <td>Second col</td>
//           </tr>
//       </tbody>
//     </table>

//   </div>

//   <div>
//     <dd>label, .label</dd>
//     <label>Here is the label</label>
//     <dd>.sublabel</dd>
//     <div class="sublabel">Here is a longer description what does this widget actually do</div>
//     <dd>input[text]</dd>
//     <input type="text" />
//     <dd>input[range]</dd>
//     <input type="range" />
//     <dd>.button_primary</dd>

//     <div class="button_primary">I am a primary button</div>
//     <dd>.button_secondary</dd>
//     <div class="button_secondary">I am a secondary button</div>
//     <dd>.button_tertiary</dd>
//     <div class="button_tertiary">I am tertiary button</div>
//     <dd>.buttons .button_tertiary</dd>
//     <div class="buttons">
//       <div class="button_tertiary">First</div>
//       <div class="button_tertiary">Second</div>
//     </div>

//   </div>

//   </div>
//   `
// }

new Vue({
  el: "#app",
  data: {
    sortedComponents,
    tags: ["2D", "3D", "Data", "Transitions", "Content", "Layout"],
    activeTag: 0
  },
  components: { ComponentRow, Colors, Styles, Utils, Faq },
  template: `
    <f-theme>

      <header>
        <div>Fachwerk is a part of <a href="https://designstem.github.io/homepage">DesignSTEM</a> education initiative.</div>
        <f-icon-github />
      </header>
  
      <div class="hero">
        <h1>Love <i>Ordnung</i>?<br>Here is Fachwerk.</h1>
        <h4 style="width: 70vw">Web framework for creating interactive learning materials. Only the <span style="color: var(--gray)">latest Chrome</span> for now.</h4>
      </div>

      <div style="padding: var(--base5); background: var(--yellow)">
        <div class="grid" style="
          --cols: 1fr 1fr 1fr;
          --gap: 1rem 2rem;
        ">
          <div>
            <h4><big>Play</big></h4>
            <div class="intro">
            Best way to get the glimpse what the framework offers is just try it out on this page. There are almost 50 components to try out - everything from 2D/3D graphics, user interaction, content creation and more.
            </div>
          </div>
          <div>
            <h4><big>Edit</big></h4>
              <div class="intro">
              We are building the web editor for easy authoring of documents. You can mix Markdown markup and Fachwerk components. The content is auto-saved to your browser.        
              </div>
          </div>
          <div>
              <h4><big>Build</big></h4>
            <div class="intro">
            Fachwerk is built on top of VueJS so to build more complex materials it is recommended to code it in Javascript, using some of the framework's components, utilities or styling.
            </div>
          </div>
          <div>
            <a href="#components" class="intro button_primary">Play with components</a>
          </div>
          <div><a href="./editor.html" class="intro button_primary">Try the editor</a>
          </div>
          <div><a href="https://github.com/designstem/template" class="intro button_primary">Get examples in Github</a>
          </div>
        </div>
      </div>

      <f-hr />

      <div style="padding: var(--base5);">
        <h1>Typograhphy & Forms</h1>
        <div class="intro">
          A case full of moving type for laying out documents.
        </div>
        <br><br>
        <Styles />
      </div>

      <f-hr />
  
      <div style="padding: var(--base5);">
        <h1>Colors</h1>
        <div class="intro">
          A compact set of useful colors, available as CSS variables, <code>var(--purple)</code> and Javascript functions <code>color('purple')</code>.
        </div>
        <br/><br/>
        <Colors />
      </div>

      <f-hr />
  
      <a id="components" style="font-weight: normal;">
        <div style="padding: var(--base5);">
          <h1>Components</h1>
          <div class="intro">
            The essential building blocks of Fachwerk. The code samples are editable ✏️ so try them out out!
          </div>
        </div>
      </a>

      <f-hr />
      
      <f-tabs
        :buttons="tags"
        v-model="activeTag"
      />

      <div style="padding: var(--base5);">
        <ComponentRow
          v-for="(c,i) in sortedComponents.map(c => Object.values(c)[0]).filter(c => c.tag == tags[activeTag])"
          :component="c"
          :nopreview="c.hasOwnProperty('Editor')"
          :key="i"
          style="margin-bottom: calc(var(--base) * 10);"
        />
      </div>

      <f-hr />
      
      <div style="padding: var(--base5);">
        <h1>Grid</h1>
        <div class="intro">
          A <a href="https://learncssgrid.com/" target="_blank">CSS grid</a> and CSS variables based grid system. The code below is editable ✏️ so try it out!
        </div>
        <br/>
        <f-fetch-data url="./contents/grid.md">
          <f-content-editor
            slot-scope="data"
            :content="data.value"
            autosave-id="grid"
          ><f-content-document slot-scope="{content}" :content="content" /></f-content-editor> 
        </f-fetch-data>
      </div>

      <f-hr />
      
      <div style="padding: var(--base5);">
        <h1>Utils</h1>
        <div class="intro">A set of Javascript helper functions. Use them like this:</div>
        <br/>

        <pre>
// Import just a single function
import { random } from 'https://designstem.github.io/framework/utils.js'

// Import all under utils namespace
import * as utils from 'https://designstem.github.io/framework/utils.js'

// Make all utils available for Vue templates
new Vue({
  methods: { ...utils }
})
        </pre>

        <Utils />
      </div>

      <f-hr />
  
      <div style="padding: var(--base5);">
        <h1>FAQ</h1>
        <Faq />
      </div>

      <div class="footer">
        <div>
          <p>
            All code is licenced under
            <a href="https://choosealicense.com/licenses/mit/" rel="licence">
              MIT licence
            </a>.
            All content is licenced under
            <br>
            <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">
              Creative Commons Attribution 4.0 International License
            </a>.
          </p>
        </div>
        <img src="./images/erasmus_logo.svg" style="width:240px" />
      </div>

    </f-theme>
  `
});
